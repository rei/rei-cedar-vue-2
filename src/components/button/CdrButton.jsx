import clsx from 'clsx';
import modifier from '../../mixins/modifier';
import fullWidth from '../../mixins/fullWidth';
import size from '../../mixins/size';
import space from '../../mixins/space';
import style from './styles/CdrButton.scss';

export default {
  name: 'CdrButton',
  mixins: [modifier, size, space, fullWidth],
  props: {
    /**
     * Controls render as button or anchor. {button, a}
     */
    tag: {
      type: String,
      default: 'button',
      validator: value => (['button', 'a'].indexOf(value) >= 0) || false,
    },
    /**
     * Sets the button type. {button, submit, reset}
     */
    type: {
      type: String,
      default: 'button',
      validator: value => (['button', 'submit', 'reset'].indexOf(value) >= 0) || false,
    },
    /**
     * Renders an icon-only button. Default slot is disabled. Overrides size and responsiveSize props.
     */
    iconOnly: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      style,
    };
  },
  computed: {
    baseClass() {
      return 'cdr-button';
    },
    defaultClass() {
      return this.modifyClassName(this.baseClass, 'primary');
    },
    buttonSizeClass() {
      return !this.iconOnly ? this.sizeClass : null;
    },
    iconClass() {
      const classes = [];

      if (this.$slots.icon && this.$slots.default) {
        /* only add class for buttons with text + icon */
        classes.push(this.modifyClassName(this.baseClass, 'has-icon'));
      }

      if (this.iconOnly) {
        classes.push(this.modifyClassName(this.baseClass, 'icon-only'));
      }
      return classes.join(' ');
    },
  },
  render() {
    const Component = this.tag;
    return (<Component
      class={clsx(this.defaultClass,
        this.style[this.baseClass],
        this.modifierClass,
        this.buttonSizeClass,
        this.fullWidthClass,
        this.iconClass,
        this.space)}
      type={this.tag === 'button' ? this.type : null}
      {...{ on: this.$listeners }}
    >
      {this.$slots.icon}
      {this.$slots.default}
    </Component>);
  },
};
