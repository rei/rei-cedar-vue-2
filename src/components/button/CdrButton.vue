<template>
  <component
    :is="tag"
    :class="[
      modifierClass,
      sizeClass,
      fullWidthClass,
      responsiveClass,
      iconClass,
    ]"
    :type="tag === 'button' ? type : null"
    @click="onClick">
    <!-- @slot for icon -->
    <slot name="icon" />
    <!-- @slot innerHTML on the inside of the button component -->
    <slot />
  </component>
</template>

<script>
import modifier from 'mixinsdir/modifier';

/**
 * Cedar 2 component for button
 *
 * CdrButton will render either a button, or an anchor that looks like a button.
 * As such, the decision to use CdrButton vs CdrAnchor should be made based on what
 * you need the rendered element to look like.
 *
 * @author [REI Software Engineering](https://rei.github.io/rei-cedar/)
 */
export default {
  name: 'CdrButton',
  mixins: [modifier],
  props: {
    /**
     * Controls render as button or anchor. {button, a}
     */
    tag: {
      type: String,
      default: 'button',
      validator: value => (['button', 'a'].indexOf(value) >= 0) || false,
    },
    /**
     * Sets the button type. {button, submit, reset}
     */
    type: {
      type: String,
      default: 'button',
      validator: value => (['button', 'submit', 'reset'].indexOf(value) >= 0) || false,
    },
    /**
     * Adds custom click actions.
     */
    onClick: {
      type: Function,
      default: () => () => null,
    },
    /**
     * Sets width to be 100%.
    */
    fullWidth: {
      type: Boolean,
      default: false,
      validator: value => typeof value === 'boolean',
    },
    /**
     * Sets a static size for the button, which scales padding and text size. {small, medium, large}
     */
    size: {
      type: String,
      default: 'medium',
      validator: value => (['small', 'medium', 'large'].indexOf(value) >= 0) || false,
    },
    /**
     * Render a specific button size at a specific breakpoint. Takes precedence over size and fullWidth.
     * Format is size@breakpoint (ex: large@sm).
     */
    responsiveSize: {
      type: Array,
    },
    /**
     * Renders an icon-only button. Default slot is disabled. Overrides size and responsiveSize props.
     */
    iconOnly: {
      type: Boolean,
      default: false,
    },
    /**
     * Renders an icon-only button with a light fill color for use on dark backgrounds.
     * iconOnly must be true.
     */
    onDark: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    baseClass() {
      return 'cdr-button';
    },
    sizeClass() {
      return !this.iconOnly ? this.modifyClassName(this.baseClass, this.size) : null;
    },
    responsiveClass() {
      const responsiveClass = [];

      if (this.responsiveSize && !this.iconOnly) {
        this.responsiveSize.forEach((val) => {
          responsiveClass.push(this.modifyClassName(this.baseClass, val));
        });
      }

      return responsiveClass.join(' ');
    },
    iconClass() {
      const classes = [];

      if (this.$slots.icon && this.$slots.default) {
        /* only add class for buttons with text + icon */
        classes.push(this.modifyClassName(this.baseClass, 'has-icon'));
      }

      if (this.iconOnly) {
        classes.push(this.modifyClassName(this.baseClass, 'icon-only'));

        if (this.onDark) {
          classes.push(this.modifyClassName(this.baseClass, 'on-dark'));
        }
      }
      return classes.join(' ');
    },
    fullWidthClass() {
      return this.fullWidth && !this.iconOnly ?
        this.modifyClassName(this.baseClass, 'full-width') : null;
    },
  },
};
</script>

<style module>
  @import 'cssdir/settings/_index.pcss';
  @import './styles/CdrButton.pcss';
</style>
