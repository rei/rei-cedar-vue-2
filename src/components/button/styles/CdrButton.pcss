/* ==========================================================================
  # CdrButton

  TOC:

    :Base Styles
      :States
    :Modifiers
      :Style variants
        :Secondary
    :Size variants
      :large
      :medium
      :small
    :full-width
    :icon-only
    :on-dark
    :Breakpoint variants
      :xs
      :sm
      :md
      :lg
========================================================================== */

.cdr-button {
  @include redwood-display-30;

  background-color: $button-background-color;
  box-shadow: $button-box-shadow;
  border: none;
  border-radius: $button-border-radius;
  color: $button-color;
  cursor: pointer;
  display: inline-flex;
  fill: $button-color;
  outline: none;
  padding: $inset-1-x-squish;
  text-align: left;
  text-decoration: none;
  text-transform: none;
  overflow: visible;

  & > :global(&__icon) {
    align-self: center;
    fill: inherit;
    margin-right: $space-quarter-x;
  }

  /* States
    ---------- */

  &:hover,
  &:active,
  &:focus {
    background-color: $button-hover-background-color;
    color: $button-hover-color;
    box-shadow: $button-hover-box-shadow;
    text-decoration: none;
    outline: none;
    outline-offset: 0;
    fill: $button-hover-color;
  }

  &:focus {
    outline: $default-outline;
  }

  &[disabled] {
    background-color: $button-disabled-background-color;
    box-shadow: $button-disabled-box-shadow;
    color: $button-disabled-color;
    cursor: not-allowed;
  }

  /* Modifier style variants
    ========================================================================== */

  /* Secondary
     ========== */

  &--secondary {
    background-color: $button-secondary-background-color;
    box-shadow: $button-secondary-box-shadow;
    color: $button-secondary-color;
    fill: $button-secondary-color;

    /* States
    ---------- */

    &:hover,
    &:active,
    &:focus {
      background-color: $button-secondary-hover-background-color;
      box-shadow: $button-secondary-box-shadow;
      color: $button-secondary-hover-color;
      fill: $button-secondary-hover-color;
    }

    &[disabled] {
      background-color: $button-secondary-disabled-background-color;
      box-shadow: $button-secondary-disabled-box-shadow;
      color: $button-secondary-disabled-color;
      cursor: not-allowed;
      fill: $button-secondary-disabled-color;
    }
  }

  /* Size variants
    ========================================================================== */

  /* Large
     ========== */
  &--large {
    @include redwood-display-40;

    padding: $inset-1-and-a-half-x-squish;

    &.cdr-button--has-icon {
      padding-left: calc($space-1-and-a-half-x - $space-quarter-x);
    }
  }

  /* Medium
     ========== */
  &--medium {
    @include redwood-display-30;

    padding: $inset-1-x-squish;

    &.cdr-button--has-icon {
      padding-left: calc($space-1-x - $space-quarter-x);
    }
  }

  /* Small
     ========== */
  &--small {
    @include redwood-display-20;

    padding: $inset-three-quarter-x-squish;

    &.cdr-button--has-icon {
      padding-left: calc($space-three-quarter-x - $space-quarter-x);
    }

    & > :global(.cdr-button__icon) {
      /* This is to correct icon size vs line height */
      @include small-button-icon();
    }
  }

  /* Full Width
     ========== */
  &--full-width {
    width: 100%;
    justify-content: center;
  }

  /* Icon Only
     ========= */

  &--icon-only {
    background: none;
    border: none;
    box-shadow: none;
    display: inline-block;
    padding: $inset-half-x;
    padding-right: calc($inset-half-x - $space-quarter-x);
    fill: $heart-of-darkness;

    &:hover,
    &:active,
    &:focus {
      background: none;
      box-shadow: none;
      fill: $heart-of-darkness;
    }

    &.cdr-button--on-dark {
      fill: $moon-shot;
    }
  }

  /* Breakpoint variants
    ========================================================================== */

  @mixin auto-icon() {
    width: $icon-size;
    height: $icon-size;
  }

  @mixin small-button-icon() {
    width: 20px;
    height: 20px;
  }

  /* @xs
    0px - 767px
    ========== */
  @media (--xs-mq-only) {
    &--small\@xs {
      @include redwood-display-20;

      padding: $inset-three-quarter-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include small-button-icon();
      }
    }

    &--medium\@xs {
      @include redwood-display-30;

      padding: $inset-1-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include auto-icon();
      }
    }

    &--large\@xs {
      @include redwood-display-40;

      padding: $inset-1-and-a-half-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include auto-icon();
      }
    }
  }

  /* @sm
    768px - 991px
    ========== */
  @media (--sm-mq-only) {
    &--small\@sm {
      @include redwood-display-20;

      padding: $inset-three-quarter-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include small-button-icon();
      }
    }

    &--medium\@sm {
      @include redwood-display-30;

      padding: $inset-1-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include auto-icon();
      }
    }

    &--large\@sm {
      @include redwood-display-40;

      padding: $inset-1-and-a-half-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include auto-icon();
      }
    }
  }

  /* @md
    992px - 1199px
    ========== */
  @media (--md-mq-only) {
    &--small\@md {
      @include redwood-display-20;

      padding: $inset-three-quarter-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include small-button-icon();
      }
    }

    &--medium\@md {
      @include redwood-display-30;

      padding: $inset-1-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include auto-icon();
      }
    }

    &--large\@md {
      @include redwood-display-40;

      padding: $inset-1-and-a-half-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include auto-icon();
      }
    }
  }

  /* @lg
    1200px and up
    ========== */

  @media (--lg-mq-only) {
    &--small\@lg {
      @include redwood-display-20;

      padding: $inset-three-quarter-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include small-button-icon();
      }
    }

    &--medium\@lg {
      @include redwood-display-30;

      padding: $inset-1-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include auto-icon();
      }
    }

    &--large\@lg {
      @include redwood-display-40;

      padding: $inset-1-and-a-half-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include auto-icon();
      }
    }
  }
}
