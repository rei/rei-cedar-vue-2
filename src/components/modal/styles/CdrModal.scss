@import '../../../css/settings/index.scss';

$modal-animation-duration: 150ms;

.cdr-modal {
  bottom: 0;
  height: 100%;
  left: 0;
  overflow-y: scroll;
  position: fixed;
  right: 0;
  top: 0;
  visibility: visible;
  z-index: 1000;

  &__overlay {
    background-color: $cdr-color-background-modal-overlay;
    backdrop-filter: blur($cdr-space-one-x);
    bottom: 0;
    left: 0;
    opacity: 1;
    position: fixed;
    right: 0;
    top: 0;
    transition: opacity $modal-animation-duration;
    z-index: 0;
  }

  &__outerWrap {
    display: flex;
    box-sizing: border-box;
    min-height: 100%;
    position: static;
    z-index: -1;
  }

  &__contentWrap {
    align-items: flex-start;
    background-color: $cdr-color-background-primary;
    display: flex;
    flex-direction: column;
    margin: auto;
    min-height: 100%;
    opacity: 1;
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    transition: opacity $modal-animation-duration $modal-animation-duration;
    width: 100%; // IE 11
    z-index: 0;
    &:focus {
      outline: 0;
    }
  }

  &__innerWrap {
    width: 100%; // IE 11
    padding: $cdr-space-one-x;
  }

  &__dialog {
    width: 100%;
  }

  &__content {
    position: relative;
  }

  &__header {
    display: flex;
    padding-bottom: $cdr-space-one-x;
  }

  &__title {
    flex: auto;
  }

  &__close-button {
    align-self: flex-start;
    border: 1px solid $cdr-color-icon-default;
    flex: none;
    margin-left: $cdr-space-half-x;
    padding: 7px; // NOTE: button should be 40x40, so drop padding by 1px
    position: relative;
    top: -2px;
    right: -4px;

    &:hover, &:focus {
      border: 1px solid $cdr-color-icon-default;
    }
  }

  &__text {
    padding: 0;
    position: relative;
  }

  &__text-content {
    overflow: auto;
    padding-right: $cdr-space-one-and-a-half-x;
    position: relative;
  }

  &__text-fade {
    position: absolute;
    bottom: 0;
    left: 0;
    height: $cdr-space-two-x;
    /* full transparency doesn't fly on safari */
    background: linear-gradient(rgba(255,255,255,0.001), rgba(255,255,255,1));
    background-attachment: scroll;
    width: 100%;
  }

  @media (min-width: $cdr-breakpoint-sm) {
    &__outerWrap {
      padding: $cdr-space-one-x;
    }

    &__innerWrap {
      padding: $cdr-space-one-and-a-half-x;
    }

    &__contentWrap {
      border-radius: $cdr-radius-softer;
      box-shadow: $cdr-prominence-lifted;
      min-height: auto;
      position: relative;
      top: auto;
      left: auto;
      right: auto;
    }

    &__dialog {
      max-width: 640px;
    }

    &__close-button {
      margin-left: $cdr-space-three-quarter-x;
      right: -8px;
      top: -5px;
    }

    &__text-content {
      padding-right: $cdr-space-two-x;
    }
  }
}

.closed {
  transition: z-index 0s $modal-animation-duration * 2,
  visibility 0s $modal-animation-duration * 2;
  visibility: hidden;
  z-index: -1;

  .cdr-modal__overlay {
    opacity: 0;
    transition: opacity $modal-animation-duration $modal-animation-duration;
  }

  .cdr-modal__contentWrap {
    opacity: 0;
    transition: opacity $modal-animation-duration;
  }
}

.cdr-modal__noscroll {
  overflow: hidden !important;
  position: fixed !important;
  width: 100%;
}
