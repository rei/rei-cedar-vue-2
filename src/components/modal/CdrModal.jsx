import debounce from 'lodash-es/debounce';
import tabbable from 'tabbable';
import clsx from 'clsx';
import {
  CdrBreakpointSm, CdrSpaceOneX, CdrSpaceTwoX,
} from '@rei/cdr-tokens/dist/js/cdr-tokens.esm';
import style from './styles/CdrModal.scss';
import onTransitionEnd from './onTransitionEnd';
import CdrButton from '../button/CdrButton';
import IconXLg from '../icon/comps/x-lg';
import CdrText from '../text/CdrText';

export default {
  name: 'CdrModal',
  components: {
    CdrButton,
    IconXLg,
    CdrText,
  },
  props: {
    opened: {
      type: Boolean,
      required: true,
    },
    label: {
      type: String,
      required: true,
    },
    showTitle: {
      type: Boolean,
      required: false,
      default: true,
    },
    ariaDescribedbBy: {
      type: String,
      required: false,
      default: null,
    },
    id: {
      type: String,
      required: false,
      default: null,
    },
    overlayClass: String,
    wrapperClass: String,
    contentClass: String,
    animationDuration: {
      type: Number,
      default: 300,
    },
  },
  data() {
    return {
      style,
      unsubscribe: null,
      keyHandler: null,
      lastActive: null,
      focusHandler: null,
      reallyClosed: !this.opened,
      isOpening: false,
      offset: null,
      headerHeight: 0,
      totalHeight: 0,
      scrollHeight: 0,
      offsetHeight: 0,
      fullscreen: false,
    };
  },
  computed: {
    dialogAttrs() {
      return {
        'aria-describedby': this.ariaDescribedBy,
        'aria-modal': 'true',
        id: this.id,
      };
    },
    verticalSpace() {
      // contentWrap vertical padding
      const fullscreen = Number(CdrSpaceTwoX);
      const windowed = Number(CdrSpaceTwoX) + Number(CdrSpaceOneX);

      return this.fullscreen
        ? fullscreen
        : windowed + fullscreen; // fullscreen, here, would account for outerWrap padding, which is the same CdrSpaceTwoX
    },
    scrollMaxHeight() {
      return this.totalHeight
        - this.headerHeight
        - this.verticalSpace;
    },
    scrolling() {
      return this.scrollHeight > this.offsetHeight;
    },
  },
  watch: {
    opened(newValue, oldValue) {
      if (!!newValue === !!oldValue) return;
      if (newValue) {
        this.handleOpened();
      } else {
        this.handleClosed();
      }
    },
  },
  mounted() {
    if (this.opened) {
      this.handleOpened();
    }

    window.addEventListener('resize', debounce(() => {
      this.measureContent();
    }, 300));
  },
  beforeDestroy() {
    if (this.unsubscribe) this.unsubscribe();
    if (this.opened) this.removeNoScroll();
    document.removeEventListener('focusin', this.focusHandler, true);
    document.removeEventListener('keydown', this.keyHandler);
  },
  methods: {
    measureContent() {
      this.$nextTick(() => {
        this.totalHeight = window.innerHeight;
        this.fullscreen = window.innerWidth < CdrBreakpointSm;
        this.headerHeight = this.$refs.header.offsetHeight;
        this.scrollHeight = this.$refs.content.scrollHeight;
        this.offsetHeight = this.$refs.content.offsetHeight;
      });
    },
    handleKeyDown({ key }) {
      switch (key) {
        case 'Escape':
        case 'Esc':
          this.onClick();
          break;
        default: break;
      }
    },
    handleFocus(e) {
      const { documentElement } = document;
      if (this.$refs.modal.contains(e.target) || !documentElement) return;

      const tabbables = tabbable(documentElement);
      const these = tabbable(this.$refs.modal);
      const nextIx = tabbables.indexOf(e.target);
      const firstModalIx = tabbables.indexOf(these[0]);
      const nextRef = nextIx < firstModalIx ? these[these.length - 1] : these[0];
      if (nextRef) nextRef.focus();

      documentElement.scrollTop = this.scrollTop;
      documentElement.scrollLeft = this.scrollLeft;
    },
    handleOpened() {
      const { activeElement } = document;
      this.addNoScroll();
      this.isOpening = true;
      this.reallyClosed = false;
      this.lastActive = activeElement;

      this.$nextTick(() => {
        if (this.$refs.modal) this.$refs.modal.focus(); // wrapped in if so testing error isn't thrown
        this.measureContent();
        this.addHandlers();

        setTimeout(() => {
          // for some reason Safari scrolls the wrapper down a bit?
          // doesn't work without setTimeout for some unknown reason
          if (this.$refs.wrapper) this.$refs.wrapper.scrollTop = 0;

          // there is a race condition for measuring overflow when modal defaults to open,
          // this seems to cover that
          this.measureContent();
        });
      });
    },
    handleClosed() {
      const { documentElement } = document;
      document.removeEventListener('keydown', this.keyHandler);
      document.removeEventListener('focusin', this.focusHandler, true);
      this.isOpening = false;

      this.unsubscribe = onTransitionEnd(
        this.$refs.wrapper,
        () => {
          if (this.isOpening) return;
          this.unsubscribe();
          this.removeNoScroll();
          this.unsubscribe = null;
          this.reallyClosed = true;

          // handle scroll-behavior: smooth
          if (documentElement) documentElement.style.scrollBehavior = 'auto';
          // restore previous scroll position
          window.scrollTo(this.offset.x, this.offset.y);
          if (documentElement) documentElement.style.scrollBehavior = '';

          if (this.lastActive) this.lastActive.focus();
        },
        this.animationDuration + 16,
      );
    },
    addNoScroll() {
      const { documentElement, body } = document;
      const offset = {
        x: window.scrollX
          || (documentElement || {}).scrollLeft
          || (body || {}).scrollLeft
          || 0,
        y: window.scrollY
          || (documentElement || {}).scrollTop
          || (body || {}).scrollTop
          || 0,
      };
      this.offset = offset;

      if (documentElement) {
        documentElement.classList.add(style['cdr-modal__noscroll']);
        // keep current scroll position manually
        documentElement.style.top = `-${offset.y}px`;
        documentElement.style.left = `-${offset.x}px`;
      }

      if (body) {
        body.classList.add(style['cdr-modal__noscroll']);
      }
    },
    removeNoScroll() {
      const { documentElement, body } = document;

      if (body) {
        body.classList.remove(style['cdr-modal__noscroll']);
      }

      if (documentElement) {
        documentElement.classList.remove(style['cdr-modal__noscroll']);
        documentElement.style.top = '';
        documentElement.style.left = '';
      }
    },
    addHandlers() {
      this.focusHandler = this.handleFocus.bind(this);
      this.keyHandler = this.handleKeyDown.bind(this);
      document.addEventListener('focusin', this.focusHandler, true);
      document.addEventListener('keydown', this.keyHandler);
    },
    onClick(e) {
      this.$emit('closed', e);
    },
  },
  render() {
    const {
      onClick,
      opened,
      label,
      wrapperClass,
      overlayClass,
      contentClass,
      reallyClosed,
    } = this;
    return (
      <div
        class={clsx(
          this.style['cdr-modal'],
          {
            [this.style.closed]: !opened,
          },
        )}
        ref="wrapper"
      >
        <div class={clsx(this.style['cdr-modal__outerWrap'], wrapperClass)}>
          <div
            aria-hidden="true"
            onClick={onClick}
            class={clsx(this.style['cdr-modal__overlay'], overlayClass)}
          />
          <div tabIndex={opened ? '0' : undefined} />
          {/*
            this is the "uncommon case of an element that captures incoming tab traversal
            for a composite widget" mentioned in the rule docs
          */}
          <div
            ref="modal"
            class={clsx(this.style['cdr-modal__contentWrap'], this.style['cdr-modal__dialog'])}
            tabIndex="-1"
            role="dialog"
            aria-modal={!!opened}
            aria-label={label}
            {...{ attrs: this.dialogAttrs }}
          >
            {this.$slots.modal || (<div
              class={clsx(this.style['cdr-modal__innerWrap'], contentClass)}
              style={reallyClosed
                ? { display: 'none' }
                : undefined
              }
            >
              <section>
                <div class={this.style['cdr-modal__content']}>
                  <div
                    class={this.style['cdr-modal__header']}
                    ref="header"
                  >
                    <div class={this.style['cdr-modal__title']}>
                      {
                        this.showTitle && this.$slots.title
                      }
                      {
                        this.showTitle && !this.$slots.title && (
                          <cdr-text
                            tag="h1"
                            modifier="heading-serif-600"
                          >
                            {this.label}
                          </cdr-text>
                        )
                      }
                    </div>
                    <cdr-button
                      class={this.style['cdr-modal__close-button']}
                      icon-only
                      with-background={true}
                      on-click={onClick}
                      aria-label="Close"
                    >
                      <IconXLg
                        slot="icon"
                        inherit-color
                      />
                    </cdr-button>
                  </div>
                  <div
                    role="document"
                    class={this.style['cdr-modal__text']}
                  >
                    <div
                      class={this.style['cdr-modal__text-content']}
                      style={ { maxHeight: `${this.scrollMaxHeight}px` } }
                      ref="content"
                      tabindex="0"
                    >
                      {this.$slots.default}
                    </div>
                    {
                      this.scrolling && (
                        <div
                        class={this.style['cdr-modal__text-fade']}
                      />
                      )
                    }
                  </div>
                </div>
              </section>
            </div>)}
          </div>
          <div tabIndex={opened ? '0' : undefined} />
        </div>
      </div>
    );
  },
};
