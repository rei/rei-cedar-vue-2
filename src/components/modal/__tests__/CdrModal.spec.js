import { shallowMount, mount } from '@vue/test-utils';
import CdrModal from 'componentdir/modal/CdrModal';
import Vue from 'vue';
// import packageJson from '../../../../package.json';
import CdrButton from 'componentdir/button/CdrButton';

describe('CdrModal.vue', () => {
  it('renders all slots, handleClosed', () => {
    const wrapper = shallowMount(CdrModal, {
      propsData: {
        opened: true,
        label: "Label is the modal title"
      },
      slots: {
        title: 'Terms and Conditions',
        stickyContentSlot: 'Sticky content',
        scrollingContentSlot: 'Main content',
        footer: 'Footer content'
      },
      attachToDocument: true,
    });
    /*
      - begins opened
      - renders all slots for screenshots
      - close modal
    */

    expect(wrapper.element).toMatchSnapshot();
    wrapper.setProps({ opened: false });
    setTimeout(() => {
      expect(wrapper.vm.reallyClosed).toBe(true);
    }, 300);
  });

  it('leaves optional slots empty, handleOpened', () => {
    const wrapper = shallowMount(CdrModal, {
      propsData: {
        opened: false,
        label: "Label is the modal title"
      },
      slots: {
        scrollingContentSlot: 'Main content',
      },
      attachToDocument: true,
    });

    wrapper.setProps({ opened: true });
    Vue.nextTick(() => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  it('handleKeyDown', () => {
    const spyOnClick = jest.fn();
    const wrapper = shallowMount(CdrModal, {
      propsData: {
        opened: true,
        label: "Label is the modal title"
      },
      slots: {
        scrollingContentSlot: 'Main content',
      },
      methods: {
        onClick: spyOnClick
      },
      attachToDocument: true,
    });

    wrapper.trigger('keydown', {
      key: 'a'
    });
    expect(spyOnClick).not.toHaveBeenCalled();

    wrapper.trigger('keydown', {
      key: 'Esc',
    });

    wrapper.trigger('keydown', {
      key: 'Escape',
    });

    expect(spyOnClick).toHaveBeenCalledTimes(2);
  });

  xit('scrolling and fullscreen snapshot', () => {
    const wrapper = shallowMount(CdrModal, {
      propsData: {
        opened: true,
        label: "Label is the modal title"
      },
      slots: {
        scrollingContentSlot: 'Main content',
      },
      attachToDocument: true,
    });

    Vue.nextTick(() => {
      wrapper.setProps({ fullscreen: true, scrollHeight: 500, offsetHeight: 400 });
      expect(wrapper.vm.scrolling).toBe(true);
      expect(wrapper.element).toMatchSnapshot();
    })
  });
});