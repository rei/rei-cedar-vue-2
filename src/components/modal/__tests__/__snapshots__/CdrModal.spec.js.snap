// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrModal.vue leaves optional slots empty, handleOpened 1`] = `
<div
  class="cdr-modal"
  role="presentation"
>
  <div
    class="cdr-modal__outerWrap"
  >
    <div
      aria-hidden="true"
      class="cdr-modal__overlay"
    />
    <div
      tabindex="0"
    />
    <div
      aria-label="Label is the modal title"
      aria-modal="true"
      class="cdr-modal__contentWrap cdr-modal__dialog"
      role="dialog"
      tabindex="-1"
    >
      <div
        class="cdr-modal__innerWrap"
        style=""
      >
        <section>
          <div
            class="cdr-modal__content"
          >
            <div
              class="cdr-modal__header"
            >
              <div
                class="cdr-modal__title"
              >
                <cdr-text-stub
                  modifier="heading-serif-600"
                  space=""
                  tag="h1"
                >
                  Label is the modal title
                </cdr-text-stub>
              </div>
              <cdr-button-stub
                aria-label="close"
                class="cdr-modal__close-button"
                icononly="true"
                id="close-modal-button"
                size="medium"
                space=""
                tag="button"
                type="button"
              >
                <iconxlg-stub
                  inheritcolor="true"
                  size="medium"
                  space=""
                />
              </cdr-button-stub>
            </div>
            <div
              class="cdr-modal__text"
              role="document"
            >
              <div
                class="cdr-modal__text-content"
                style="max-height: 688px;"
                tabindex="0"
              >
                Main content
              </div>
            </div>
          </div>
        </section>
      </div>
    </div>
    <div
      tabindex="0"
    />
  </div>
</div>
`;

exports[`CdrModal.vue renders all slots, handleClosed 1`] = `
<div
  class="cdr-modal"
  role="presentation"
>
  <div
    class="cdr-modal__outerWrap"
  >
    <div
      aria-hidden="true"
      class="cdr-modal__overlay"
    />
    <div
      tabindex="0"
    />
    <div
      aria-label="Label is the modal title"
      aria-modal="true"
      class="cdr-modal__contentWrap cdr-modal__dialog"
      role="dialog"
      tabindex="-1"
    >
      <div
        class="cdr-modal__innerWrap"
      >
        <section>
          <div
            class="cdr-modal__content"
          >
            <div
              class="cdr-modal__header"
            >
              <div
                class="cdr-modal__title"
              >
                Terms and Conditions
              </div>
              <cdr-button-stub
                aria-label="close"
                class="cdr-modal__close-button"
                icononly="true"
                id="close-modal-button"
                size="medium"
                space=""
                tag="button"
                type="button"
              >
                <iconxlg-stub
                  inheritcolor="true"
                  size="medium"
                  space=""
                />
              </cdr-button-stub>
            </div>
            <div
              class="cdr-modal__sticky-content"
              tabindex="0"
            >
              Sticky content
            </div>
            <div
              class="cdr-modal__text"
              role="document"
            >
              <div
                class="cdr-modal__text-content"
                style="max-height: -80px;"
                tabindex="0"
              >
                Main content
              </div>
              <div
                class="cdr-modal__footer"
              >
                Footer content
              </div>
            </div>
          </div>
        </section>
      </div>
    </div>
    <div
      tabindex="0"
    />
  </div>
</div>
`;

exports[`CdrModal.vue scrolling and fullscreen snapshot 1`] = `
<div
  class="cdr-modal"
  role="presentation"
>
  <div
    class="cdr-modal__outerWrap"
  >
    <div
      aria-hidden="true"
      class="cdr-modal__overlay"
    />
    <div
      tabindex="0"
    />
    <div
      aria-label="Label is the modal title"
      aria-modal="true"
      class="cdr-modal__contentWrap cdr-modal__dialog"
      role="dialog"
      tabindex="-1"
    >
      <div
        class="cdr-modal__innerWrap"
      >
        <section>
          <div
            class="cdr-modal__content"
          >
            <div
              class="cdr-modal__header"
            >
              <div
                class="cdr-modal__title"
              >
                <cdr-text-stub
                  modifier="heading-serif-600"
                  space=""
                  tag="h1"
                >
                  Label is the modal title
                </cdr-text-stub>
              </div>
              <cdr-button-stub
                aria-label="close"
                class="cdr-modal__close-button"
                icononly="true"
                id="close-modal-button"
                size="medium"
                space=""
                tag="button"
                type="button"
              >
                <iconxlg-stub
                  inheritcolor="true"
                  size="medium"
                  space=""
                />
              </cdr-button-stub>
            </div>
            <div
              class="cdr-modal__text"
              role="document"
            >
              <div
                class="cdr-modal__text-content"
                style="max-height: 688px;"
                tabindex="0"
              >
                Main content
              </div>
            </div>
          </div>
        </section>
      </div>
    </div>
    <div
      tabindex="0"
    />
  </div>
</div>
`;
